test/lib/jdk/test/lib/cds/CDSArchiveUtils.java

```diff
@@ -0,0 +1,211 @@
+ * questions.
+ */
+package jdk.test.lib.cds;
+
+import java.io.IOException;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import jdk.test.lib.Utils;
+
+import java.nio.file.Files;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.channels.FileChannel;
+import static java.nio.file.StandardCopyOption.REPLACE_EXISTING;
+import java.nio.file.StandardOpenOption;
+import static java.nio.file.StandardOpenOption.READ;
+import static java.nio.file.StandardOpenOption.WRITE;
+
+import jdk.test.lib.cds.CDSTestUtils;
+import sun.hotspot.WhiteBox;
+
+// This class performs operations on shared archive file
+public class CDSArchiveUtils {
+    // wb must be assigned first to use the functions
+    public static WhiteBox wb;
+    public static int offsetMagic;           // CDSFileMapHeaderBase::_magic
+    public static int offsetVersion;         // CDSFileMapHeaderBase::_version
+    public static int offsetJvmIdent;        // FileMapHeader::_jvm_ident
+    public static int spOffsetCrc;           // CDSFileMapRegion::_crc
+    private static int fileHeaderSize = -1;  // total size of header, aligned with pageSize
+    public static int cdsFileMapRegionSize;  // size of CDSFileMapRegion
+    public static int spOffset;              // offset of CDSFileMapRegion
+    public static int spUsedOffset;          // offset of CDSFileMapRegion::_used
+    public static int sizetSize;             // size of size_t
+    public static int intSize;               // size of int
+    public static int pageSize;              // page size
+    public static long alignment;            // MetaspaceShared::core_region_alignment
+
+    public static void initialize(WhiteBox box) throws Exception {
+        wb = box;
+        offsetMagic = wb.getOffsetForName("FileMapHeader::_magic");
+        offsetVersion = wb.getOffsetForName("FileMapHeader::_version");
+        offsetJvmIdent = wb.getOffsetForName("FileMapHeader::_jvm_ident");
+        spOffsetCrc = wb.getOffsetForName("CDSFileMapRegion::_crc");
+        try {
+            int nonExistOffset = wb.getOffsetForName("FileMapHeader::_non_exist_offset");
+            System.exit(-1); // should fail
+        } catch (Exception e) {
+            // success
+        }
+
+        spOffset = wb.getOffsetForName("FileMapHeader::_space[0]") - offsetMagic;
+        spUsedOffset = wb.getOffsetForName("CDSFileMapRegion::_used") - spOffsetCrc;
+        sizetSize = wb.getOffsetForName("size_t_size");
+        intSize = wb.getOffsetForName("int_size");
+        cdsFileMapRegionSize  = wb.getOffsetForName("CDSFileMapRegion_size");
+        pageSize = wb.getVMPageSize();
+        alignment = wb.metaspaceSharedRegionAlignment();
+        // fileHeaderSize may not be available
+        // fileHeaderSize = (int)alignUpWithPageSize(wb.getOffsetForName("file_header_size"));
+    }
```

Comment:
This should always be available. `file_header_size` is a constant value (`sizeof(FileMapHeader)`) stored inside cdsoffsets.cpp. It doesn't depend on the contents of the archive file.

Reply:
This is same as original code, I will investigate more of this part. Thanks.